source/common/runtime/runtime_impl.h

```diff
@@ -107,91 +112,63 @@ class SnapshotImpl : public Snapshot,
+
+  // Create a new Snapshot
+  virtual std::unique_ptr<SnapshotImpl> createNewSnapshot();
+  // Load a new Snapshot into TLS
+  void loadNewSnapshot();
 
-  Filesystem::WatcherPtr watcher_;
-  ThreadLocal::SlotPtr tls_;
   RandomGenerator& generator_;
-  std::string root_path_;
-  std::string override_path_;
-  std::shared_ptr<SnapshotImpl> current_snapshot_;
-  RuntimeStats stats_;
-  Api::OsSysCallsPtr os_sys_calls_;
+  std::unordered_map<std::string, std::string> values_;
+
+private:
+  ThreadLocal::SlotPtr tls_;
 };
 
 /**
- * Null implementation of runtime if another runtime source is not configured.
+ * Extension of LoaderImpl that watches a symlink for swapping and loads a specified subdirectory
+ * from disk. Values added via mergeValues() are secondary to those loaded from disk.
  */
-class NullLoaderImpl : public Loader {
+class DiskBackedLoaderImpl : public LoaderImpl {
 public:
-  NullLoaderImpl(RandomGenerator& generator) : snapshot_(generator) {}
-
-  // Runtime::Loader
-  Snapshot& snapshot() override { return snapshot_; }
+  DiskBackedLoaderImpl(Event::Dispatcher& dispatcher, ThreadLocal::SlotAllocator& tls,
+                       const std::string& root_symlink_path, const std::string& subdir,
+                       const std::string& override_dir, Stats::Store& store,
+                       RandomGenerator& generator, Api::OsSysCallsPtr os_sys_calls);
 
 private:
-  struct NullSnapshotImpl : public Snapshot {
-    NullSnapshotImpl(RandomGenerator& generator) : generator_(generator) {}
-
-    // Runtime::Snapshot
-    bool featureEnabled(const std::string&, uint64_t default_value, uint64_t random_value,
-                        uint64_t num_buckets) const override {
-      return random_value % num_buckets < std::min(default_value, num_buckets);
-    }
-
-    bool featureEnabled(const std::string& key, uint64_t default_value) const override {
-      if (default_value == 0) {
-        return false;
-      } else if (default_value == 100) {
-        return true;
-      } else {
-        return featureEnabled(key, default_value, generator_.random());
-      }
-    }
-
-    bool featureEnabled(const std::string& key, uint64_t default_value,
-                        uint64_t random_value) const override {
-      return featureEnabled(key, default_value, random_value, 100);
-    }
-
-    const std::string& get(const std::string&) const override { return EMPTY_STRING; }
-
-    uint64_t getInteger(const std::string&, uint64_t default_value) const override {
-      return default_value;
-    }
-
-    const std::unordered_map<std::string, const Snapshot::Entry>& getAll() const override {
-      return values_;
-    }
-
-    RandomGenerator& generator_;
-    std::unordered_map<std::string, const Snapshot::Entry> values_;
-  };
+  std::unique_ptr<SnapshotImpl> createNewSnapshot() override;
+  RuntimeStats generateStats(Stats::Store& store);
 
-  NullSnapshotImpl snapshot_;
+  Filesystem::WatcherPtr watcher_;
+  std::string root_path_;
```

Comment:
I know you copied tis code, but can you constify `root_path_` and `override_path_`

Reply:
thanks good catch